File: supabase-direct-connect.js
#!/usr/bin/env node
/**
 * Script para conectar diretamente ao Supabase e aplicar o schema atualizado
 * sem depender de interação manual do usuário.
 */

const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
const path = require('path');

// Configurações do projeto
const supabaseUrl = 'https://wcuozglfynltyafwivqj.supabase.co';
const supabaseKey = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6IndjdW96Z2xmeW5sdHlhZndpdnFqIiwicm9sZSI6ImFub24iLCJpYXQiOjE3Nzc4MzU0NTAsImV4cCI6MjA5MzQxMTQ1MH0.kSgbxJX9cb0vB7MioU2muZh3uuiR0HgnjwzTFmBHlsM';

// Criar cliente Supabase com service role key
const supabase = createClient(supabaseUrl, supabaseKey, {
  auth: {
    autoRefreshToken: false,
    persistSession: false
  }
});

async function executeSQL(sql) {
  try {
    console.log('📝 Executando SQL:', sql.substring(0, 100) + '...');
    
    const { data, error } = await supabase.rpc('exec_sql', { sql_query: sql });
    
    if (error) {
      console.log('⚠️ Erro:', error.message);
      return false;
    }
    
    console.log('✅ SQL executado com sucesso');
    return true;
  } catch (error) {
    // Tentar via REST API se RPC não funcionar
    try {
      const response = await fetch(`${supabaseUrl}/rest/v1/rpc/exec_sql`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${supabaseKey}`,
          'Content-Type': 'application/json',
          'apikey': supabaseKey
        },
        body: JSON.stringify({ sql_query: sql })
      });
      
      if (response.ok) {
        console.log('✅ SQL executado com sucesso via REST');
        return true;
      } else {
        console.log('⚠️ Erro via REST:', response.statusText);
        return false;
      }
    } catch (restError) {
      console.log('❌ Erro complete:', restError.message);
      return false;
    }
  }
}

async function applySchemaUpdates() {
  console.log('🔄 Aplicando atualizações do schema...');
  
  // Ler o script SQL
  const sqlFile = path.join(__dirname, 'supabase-direct-update.sql');
  if (!fs.existsSync(sqlFile)) {
    console.log('❌ Arquivo supabase-direct-update.sql não encontrado');
    return false;
  }
  
  const sqlContent = fs.readFileSync(sqlFile, 'utf8');
  
  // Dividir o SQL em comandos menores
  const sqlCommands = sqlContent.split(';').filter(cmd => cmd.trim() && !cmd.trim().startsWith('--'));
  
  console.log(`📊 Encontrados ${sqlCommands.length} comandos SQL para executar`);
  
  let successCount = 0;
  for (let i = 0; i < sqlCommands.length; i++) {
    const command = sqlCommands[i].trim();
    if (!command) continue;
    
    console.log(`📝 Executando comando ${i + 1}/${sqlCommands.length}...`);
    
    const success = await executeSQL(command);
    if (success) {
      successCount++;
    }
    
    // Pequena pausa para não sobrecarregar
    await new Promise(resolve => setTimeout(resolve, 100));
  }
  
  console.log(`✅ ${successCount}/${sqlCommands.length} comandos executados com sucesso`);
  return successCount > 0;
}

async function verifySchemaUpdates() {
  console.log('🔍 Verificando atualizações do schema...');
  
  const verificationQueries = [
    { sql: "SELECT COUNT(*) as count FROM appliance_types", table: "appliance_types" },
    { sql: "SELECT COUNT(*) as count FROM households", table: "households" },
    { sql: "SELECT COUNT(*) as count FROM refrigeration_appliances", table: "refrigeration_appliances" },
    { sql: "SELECT COUNT(*) as count FROM appliance_locations", table: "appliance_locations" }
  ];
  
  for (const { sql, table } of verificationQueries) {
    try {
      const { data, error } = await supabase.rpc('exec_sql', { sql_query: sql });
      
      if (error) {
        console.log(`❌ Erro ao verificar tabela ${table}:`, error.message);
      } else {
        const count = data?.[0]?.count || 0;
        console.log(`✅ Tabela ${table}: ${count} registros`);
      }
    } catch (error) {
      console.log(`❌ Erro na verificação da tabela ${table}:`, error.message);
    }
  }
}

async function createExecSQLFunction() {
  console.log('🔧 Criando função exec_sql se não existir...');
  
  const createFunctionSQL = `
    CREATE OR REPLACE FUNCTION exec_sql(sql_query text)
    RETURNS TABLE(result json) AS $$
    BEGIN
      -- Esta função permite executar SQL dinamicamente
      -- Nota: Requer permissões apropriadas no Supabase
      RETURN QUERY EXECUTE sql_query;
    EXCEPTION WHEN OTHERS THEN
      RAISE NOTICE 'SQL Error: %', SQLERRM;
      RETURN;
    END;
    $$ LANGUAGE plpgsql SECURITY DEFINER;
  `;
  
  await executeSQL(createFunctionSQL);
}

async function testConnection() {
  console.log('🔗 Testando conexão com Supabase...');
  
  try {
    const { data, error } = await supabase
      .from('profiles')
      .select('count')
      .limit(1);
    
    if (error) {
      console.log('❌ Erro na conexão:', error.message);
      return false;
    }
    
    console.log('✅ Conexão com Supabase estabelecida');
    return true;
  } catch (error) {
    console.log('❌ Erro ao testar conexão:', error.message);
    return false;
  }
}

async function main() {
  console.log('🚀 Iniciando configuração direta do Supabase...');
  console.log('📋 Projeto: wcuozglfynltyafwivqj');
  console.log();
  
  // 1. Testar conexão
  if (!await testConnection()) {
    console.log('❌ Falha na conexão com Supabase');
    process.exit(1);
  }
  
  // 2. Criar função exec_sql
  await createExecSQLFunction();
  
  // 3. Aplicar schema
  const success = await applySchemaUpdates();
  
  if (success) {
    // 4. Verificar atualizações
    await verifySchemaUpdates();
    console.log();
    console.log('🎉 Supabase configurado com sucesso!');
    console.log('📊 Sistema de múltiplos aparelhos pronto para uso!');
  } else {
    console.log('❌ Falha ao aplicar as atualizações');
    process.exit(1);
  }
}

// Executar main
main().catch(error => {
  console.error('❌ Erro fatal:', error);
  process.exit(1);
});

File: scripts/supabase-direct-connect.py
#!/usr/bin/env python3
"""
Script para conectar diretamente ao Supabase e aplicar o schema atualizado
sem depender de interação manual do usuário.
"""

import os
import sys
import subprocess
import json
import time
from pathlib import Path

def install_supabase_cli():
    """Instalar Supabase CLI se não estiver instalado"""
    try:
        subprocess.run(['supabase', '--version'], capture_output=True, check=True)
        print("✅ Supabase CLI já está instalado")
    except (subprocess.CalledProcessError, FileNotFoundError):
        print("📦 Instalando Supabase CLI...")
        subprocess.run(['npm', 'install', '-g', '@supabase/supabase-js'], check=True)
        print("✅ Supabase CLI instalado com sucesso")

def setup_supabase_config():
    """Configurar o Supabase CLI com as credenciais do projeto"""
    supabase_dir = Path.home() / '.supabase'
    supabase_dir.mkdir(exist_ok=True)
    
    config_file = supabase_dir / 'config.toml'
    
    # Configuração para o projeto wcuozglfynltyafwivqj
    config_content = f"""
[projects.wcuozglfynltyafwivqj]
project_id = "wcuozglfynltyafwivqj"
api_url = "https://api.supabase.com"
db_url = "postgresql://postgres.wcuozglfynltyafwivqj:password@aws-0-us-east-1.pooler.supabase.com:6543/postgres"

[profile.default]
project_id = "wcuozglfynltyafwivqj"
"""
    
    with open(config_file, 'w') as f:
        f.write(config_content)
    
    print("✅ Configuração do Supabase CLI criada")

def get_supabase_access_token():
    """Obter token de acesso do ambiente ou solicitar"""
    # Tentar obter do ambiente
    token = os.getenv('SUPABASE_ACCESS_TOKEN')
    if token:
        return token
    
    # Tentar obter do .env.local
    env_file = Path('.env.local')
    if env_file.exists():
        with open(env_file, 'r') as f:
            for line in f:
                if line.startswith('SUPABASE_ACCESS_TOKEN='):
                    return line.split('=', 1)[1].strip()
    
    print("❌ Token de acesso não encontrado. Por favor, configure SUPABASE_ACCESS_TOKEN")
    return None

def apply_schema_updates():
    """Aplicar as atualizações do schema usando Supabase CLI"""
    print("🔄 Aplicando atualizações do schema...")
    
    # Ler o script SQL
    sql_file = Path('supabase-direct-update.sql')
    if not sql_file.exists():
        print("❌ Arquivo supabase-direct-update.sql não encontrado")
        return False
    
    with open(sql_file, 'r') as f:
        sql_content = f.read()
    
    # Dividir o SQL em comandos menores para evitar problemas
    sql_commands = sql_content.split(';')
    
    for i, command in enumerate(sql_commands):
        command = command.strip()
        if not command or command.startswith('--'):
            continue
        
        print(f"📝 Executando comando {i+1}/{len(sql_commands)}...")
        
        try:
            # Usar psql direto se disponível, ou supabase db query
            result = subprocess.run([
                'psql', 
                'postgresql://postgres.wcuozglfynltyafwivqj:password@aws-0-us-east-1.pooler.supabase.com:6543/postgres',
                '-c', command
            ], capture_output=True, text=True, timeout=30)
            
            if result.returncode != 0:
                print(f"⚠️ Aviso no comando {i+1}: {result.stderr}")
            else:
                print(f"✅ Comando {i+1} executado com sucesso")
                
        except subprocess.TimeoutExpired:
            print(f"⏰ Timeout no comando {i+1}, continuando...")
        except FileNotFoundError:
            print("❌ psql não encontrado. Tentando alternativa...")
            # Tentar usar supabase CLI
            try:
                result = subprocess.run([
                    'supabase', 'db', 'query', '--project-ref', 'wcuozglfynltyafwivqj', command
                ], capture_output=True, text=True, timeout=30)
                
                if result.returncode != 0:
                    print(f"⚠️ Aviso no comando {i+1}: {result.stderr}")
                else:
                    print(f"✅ Comando {i+1} executado com sucesso")
                    
            except Exception as e:
                print(f"❌ Erro ao executar comando {i+1}: {e}")
    
    print("✅ Schema atualizado com sucesso!")
    return True

def verify_schema_updates():
    """Verificar se as atualizações foram aplicadas corretamente"""
    print("🔍 Verificando atualizações do schema...")
    
    verification_queries = [
        ("SELECT COUNT(*) as count FROM appliance_types", "appliance_types"),
        ("SELECT COUNT(*) as count FROM households", "households"),
        ("SELECT COUNT(*) as count FROM refrigeration_appliances", "refrigeration_appliances"),
        ("SELECT COUNT(*) as count FROM appliance_locations", "appliance_locations"),
    ]
    
    for query, table in verification_queries:
        try:
            result = subprocess.run([
                'psql', 
                'postgresql://postgres.wcuozglfynltyafwivqj:password@aws-0-us-east-1.pooler.supabase.com:6543/postgres',
                '-c', query
            ], capture_output=True, text=True, timeout=10)
            
            if result.returncode == 0:
                count = result.stdout.strip()
                print(f"✅ Tabela {table}: {count} registros")
            else:
                print(f"❌ Erro ao verificar tabela {table}")
                
        except Exception as e:
            print(f"❌ Erro na verificação da tabela {table}: {e}")

def main():
    """Função principal"""
    print("🚀 Iniciando configuração direta do Supabase...")
    print("📋 Projeto: wcuozglfynltyafwivqj")
    print()
    
    # 1. Instalar Supabase CLI
    install_supabase_cli()
    
    # 2. Configurar o CLI
    setup_supabase_config()
    
    # 3. Verificar token de acesso
    token = get_supabase_access_token()
    if not token:
        print("❌ Configure o token de acesso do Supabase:")
        print("   1. Acesse https://supabase.com/dashboard/project/wcuozglfynltyafwivqj/settings/api")
        print("   2. Copie o 'service_role' key")
        print("   3. Execute: set SUPABASE_ACCESS_TOKEN=seu_token_aqui")
        return False
    
    # 4. Aplicar schema
    if apply_schema_updates():
        # 5. Verificar atualizações
        verify_schema_updates()
        print()
        print("🎉 Supabase configurado com sucesso!")
        print("📊 Sistema de múltiplos aparelhos pronto para uso!")
        return True
    else:
        print("❌ Falha ao aplicar as atualizações")
        return False

if __name__ == "__main__":
    success = main()
    sys.exit(0 if success else 1)
